tools/codegen/gen.py

```diff
@@ -348,31 +348,91 @@ def go(f: NativeFunction) -> Optional[str]:
             return None
 
         name = cpp.name(f.func)
-
         cpp_returns_type = cpp.returns_type(f.func.returns)
-        cpp_args = cpp.arguments(f.func)
-        cpp_args_str = ', '.join(map(str, cpp_args))
+
+        # scattered_cpp_args: ['dtype', 'layout', 'device', 'pin_memory']
+        # gathered_cpp_args: ['options']
+        scattered_cpp_args = cpp.scattered_cpp_arguments(f.func)
+        gathered_cpp_args = cpp.gathered_cpp_arguments(f.func)
+        scattered_cpp_args_str = ', '.join(map(str, scattered_cpp_args))
+        gathered_cpp_args_str = ', '.join(map(str, gathered_cpp_args))
+        scattered_cpp_args_str_no_default = ', '.join(map(lambda a: a.str_no_default(), scattered_cpp_args))
+        gathered_cpp_args_str_no_default = ', '.join(map(lambda a: a.str_no_default(), gathered_cpp_args))
+
+        # scattered_dispatcher_expr:  ['dtype', 'layout', 'device', 'pin_memory']
+        # gathered_dispatcher_expr:   ['options']
+        # scattering_dispatcher_expr: ['options.dtype_opt()', 'options.layout_opt)', ...]
+        # gathering_dispatcher_expr:  ['TensorOptions().dtype(dtype).layout(layout)...']
+        # TODO gathered_dispatcher_expr and gathering_dispatcher_expr are only needed for non-c10-full ops,
+        #      remove later.
+        scattered_dispatcher_exprs = dispatcher.cpparguments_exprs(scattered_cpp_args, process_tensoroptions=None)
+        gathered_dispatcher_exprs = dispatcher.cpparguments_exprs(gathered_cpp_args, process_tensoroptions=None)
+        scattering_dispatcher_exprs = dispatcher.cpparguments_exprs(gathered_cpp_args, process_tensoroptions='scatter')
+        gathering_dispatcher_exprs = dispatcher.cpparguments_exprs(gathered_cpp_args, process_tensoroptions='gather')
+
+        scattering_dispatcher_exprs_str = ', '.join(map(lambda a: a.expr, scattering_dispatcher_exprs))
+        gathering_dispatcher_exprs_str = ', '.join(map(lambda a: a.expr, gathering_dispatcher_exprs))
 
         if target is Target.DECLARATION:
-            return f"CAFFE2_API {cpp_returns_type} {name}({cpp_args_str});"
+            if gathered_cpp_args_str == scattered_cpp_args_str:
```

Comment:
Hmmm there's definitely a better way to test for the presence/absense of tensor options :P